### PR TITLE
Ingawei/market tiers

### PR DESCRIPTION
**Pull request by ingawei**
image
Please pay extra attention to the backend stuff!

### CODE REVIEW

File: backend/api/src/create-market.ts [hunk 1]
```diff
@@ -625,4 +646,30 @@ async function generateAntes(
...
+    )
+
+    await insertLiquidity(tx, newLiquidityProvision)
+
+    contractRef.update({
+      subsidyPool: FieldValue.increment(drizzledAmount),
+      totalLiquidity: FieldValue.increment(drizzledAmount),
+    })
+  })
+}
```
Comment (jahooma) [inline backend/api/src/create-market.ts:674]: I think this is right for binary, but for multiple choice, we want to split the drizzled amount evenly among answers, insert a liquidity doc for each, and then update the answer doc (`contracts/[contractId]/answersCpmm/[answerId]`) with subsidyPool and totalLiquidity.

**Review by jahooma — COMMENTED** [review]
Thanks Inga! This is really cool. Users will like this change.

I left a few comments for you to address.

Also, Stephen and I are still debating, but I think it's possible we will want another lower tier for markets: https://manifold.markets/JamesGrugett/will-manifold-add-a-cheaper-tier-fo

I think this PR should go ahead without the lower tier, but you might want to think about how a lower tier could be incorporated (and also happy to hear your thoughts on this question).

File: backend/api/src/create-market.ts [hunk 2]
```diff
@@ -150,13 +156,15 @@ export async function createMarketHelper(body: Body, auth: AuthedUser) {
...
+const totalMarketCost = marketTier ? getTieredCost(ante, marketTier, outcomeType): ante
+
   const contract = await pg.tx(async (tx) => {
     const user = await getUser(userId, tx)
     if (!user) throw new APIError(401, 'Your account was not found')
     if (user.isBannedFromPosting) throw new APIError(403, 'You are banned')
 
-    if (ante > user.balance)
-      throw new APIError(403, `Balance must be at least ${ante}.`)
+    if (totalMarketCost > user.balance)
```
[the remark this answers is not in the document]
Comment (ingawei) [reply backend/api/src/create-market.ts:166]: I think that's the runTxn part in line 652

File: backend/api/src/create-market.ts [hunk 3]
```diff
@@ -150,13 +156,15 @@ export async function createMarketHelper(body: Body, auth: AuthedUser) {
 
   const pg = createSupabaseDirectClient()
 
+const totalMarketCost = marketTier ? getTieredCost(ante, marketTier, outcomeType): ante
```
[the remark this answers is not in the document]
Comment (ingawei) [reply backend/api/src/create-market.ts:159]: Seems like the ante isn't actually used at all as an input, so I think I'm just gonna keep the tier and then API users can just not put in ante and have marketTier determine the price.

File: backend/api/src/create-market.ts (hunk 1 above)
Comment (ingawei) [reply backend/api/src/create-market.ts:674]: Is there a precedent for this? I just reused whatever was in the add subsidy code, does that not do that?

File: backend/api/src/create-market.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jahooma) [reply backend/api/src/create-market.ts:155]: Ok, nice, yeah that makes sense

File: backend/api/src/create-market.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jahooma) [reply backend/api/src/create-market.ts:162]: Ah, I see, I guess that works!

File: backend/api/src/create-market.ts (hunk 1 above)
Comment (jahooma) [reply backend/api/src/create-market.ts]: Oh, I guess either works for now, so this is fine.

If this is exactly the same functionality as in addContractLiquidity, you might want to extract a helper function that is used in both places. Would make it much easier to update both in the future.

Comment (ingawei) [reply backend/api/src/create-market.ts]: Yeah I thought about that but most of the stuff is already extracted into a function and it's dissimilar enough that I think it just wouldn't make sense

**Review by jahooma — APPROVED** [review]

**Pull request merged**